test/libsolidity/GasTest.cpp

```diff
@@ -0,0 +1,237 @@
+			if (*it++ != '+')
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected \"+\"-"));
+			skipWhitespace(it, line.end());
+			m_creationCost.codeDepositCost = parseGasCost(it, line.end());
+			skipWhitespace(it, line.end());
+			if (*it++ != '=')
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected \"+\"-"));
+			skipWhitespace(it, line.end());
+			m_creationCost.totalCost = parseGasCost(it, line.end());
+		}
+		else if (line == "// external:")
+			currentKind = &m_externalFunctionCosts;
+		else if (line == "// internal:")
+			currentKind = &m_internalFunctionCosts;
+		else if (!currentKind)
+			BOOST_THROW_EXCEPTION(runtime_error("No function kind specified. Expected \"external:\" or \"internal:\"."));
+		else
+		{
+			if (!boost::starts_with(line, "// "))
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected \"// \"."));
+			auto it = line.begin() + 3;
+			skipWhitespace(it, line.end());
+			auto functionNameBegin = it;
+			while (it != line.end() && *it != ':')
+				++it;
+			std::string functionName(functionNameBegin, it);
+			if (functionName == "fallback")
+				functionName.clear();
+			expect(it, line.end(), ':');
+			skipWhitespace(it, line.end());
+			if (it == line.end())
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected gas cost."));
+			(*currentKind)[functionName] = std::string(it, line.end());
+		}
+}
+
+void GasTest::printUpdatedExpectations(std::ostream& _stream, std::string const& _linePrefix) const
+{
+	Json::Value estimates = compiler().gasEstimates(compiler().lastContractName());
+	_stream << _linePrefix
+		<< "creation: "
+		<< estimates["creation"]["executionCost"].asString()
+		<< " + "
+		<< estimates["creation"]["codeDepositCost"].asString()
+		<< " = "
+		<< estimates["creation"]["totalCost"].asString()
+		<< std::endl;
+
+	for (auto kind: {"external", "internal"})
+		if (estimates[kind])
+		{
+			_stream << _linePrefix << kind << ":" << std::endl;
+			for (auto it = estimates[kind].begin(); it != estimates[kind].end(); ++it)
+			{
+				_stream << _linePrefix << "  ";
+				if (it.key().asString().empty())
+					_stream << "fallback";
+				else
+					_stream << it.key().asString();
+				_stream << ": " << it->asString() << std::endl;
+			}
+		}
+}
+
+
+bool GasTest::run(ostream& _stream, string const& _linePrefix, bool _formatted)
+{
+	string const versionPragma = "pragma solidity >=0.0;\n";
+	compiler().reset();
+	OptimiserSettings settings = m_optimise ? OptimiserSettings::standard() : OptimiserSettings::minimal();
+	if (m_optimiseYul)
+	{
+		settings.runYulOptimiser = m_optimise;
+		settings.optimizeStackAllocation = m_optimise;
+	}
+	settings.expectedExecutionsPerDeployment = m_optimiseRuns;
+	compiler().setOptimiserSettings(settings);
+	compiler().setSources({{"", versionPragma + m_source}});
+
+	if (!compiler().parseAndAnalyze() || !compiler().compile())
+	{
+		SourceReferenceFormatterHuman formatter(cerr, _formatted);
```

Comment:
Outputting to ``cerr`` here is not perfect - but we can't just output to ``_stream`` and return ``false`` without throwing, because ``isoltest`` needs to know that it can't update the expectations.

The better general solution for this would be to let ``run`` return an ``Success, Error, Fatal`` enum instead of a boolean, so that ``isoltest`` can respond accordingly - I could do that in this PR or a separate one...

Reply:
I think I'll just add another commit doing that.

Reply:
Did that - quite a few of the other kinds of test cases would actually have already needed this anyways.

Reply:
You could also include the message in the exception text.